addons/hr_attendance/tests/test_hr_attendance_undertime.py

```diff
@@ -0,0 +1,479 @@
+        })
+        self.assertAlmostEqual(m_attendance_1.overtime_hours, -4, 2)
+
+        m_attendance_2 = self.env['hr.attendance'].create({
+            'employee_id': self.employee.id,
+            'check_in': datetime(2023, 1, 3, 13, 0),
+            'check_out': datetime(2023, 1, 3, 17, 0),
+        })
+        self.assertAlmostEqual(m_attendance_1.overtime_hours, 0, 2)
+        self.assertAlmostEqual(m_attendance_2.overtime_hours, 0, 2)
+
+        m_attendance_3 = self.env['hr.attendance'].create({
+            'employee_id': self.employee.id,
+            'check_in': datetime(2023, 1, 3, 18, 0),
+            'check_out': datetime(2023, 1, 3, 19, 0),
+        })
+        self.assertAlmostEqual(m_attendance_1.overtime_hours, 0, 2)
+        self.assertAlmostEqual(m_attendance_2.overtime_hours, 0, 2)
+        self.assertAlmostEqual(m_attendance_3.overtime_hours, 1, 2)
+
+        overtime_2 = self.env['hr.attendance.overtime.line'].search([('employee_id', '=', self.employee.id),
+                                                                ('date', '=', datetime(2023, 1, 3))])
+        # Total overtime for that day : 5 hours
+        self.assertEqual(len(overtime_2), 1, "Only one overtime record should be created for that day.")
+        self.assertEqual(overtime_2.duration, 1)
+
+        # Attendance Modification case
+
+        m_attendance_3.write({
+            'check_out': datetime(2023, 1, 3, 20, 00),
+        })
+        self.assertAlmostEqual(m_attendance_3.overtime_hours, 2, 2)
+
+        # Deleting previous attendances should update correctly the overtime hours in other attendances
+        m_attendance_2.unlink()
+        self.assertAlmostEqual(m_attendance_3.overtime_hours, -2, 2)
+
+    def test_undertime_across_days_timezones(self):
+        self.europe_employee.ruleset_id = self.ruleset
+
+        early_attendance = self.env['hr.attendance'].create({
+            'employee_id': self.europe_employee.id,
+            'check_in': datetime(2024, 5, 27, 23, 30),
+            'check_out': datetime(2024, 5, 28, 4, 30),
+        })
+        self.assertAlmostEqual(early_attendance.overtime_hours, -3, 2)
+
+        # Total overtime for that day : -1 hours
+        overtime_record = early_attendance.linked_overtime_ids
+        self.assertAlmostEqual(overtime_record.duration, -3, 2)
+
+        # Check that the calendar's timezones take priority and that overtimes and attendances dates are consistent
+        self.europe_employee.resource_calendar_id.tz = 'America/New_York'
+
+        early_attendance2 = self.env['hr.attendance'].create({
+            'employee_id': self.europe_employee.id,
+            'check_in': datetime(2024, 5, 30, 3, 0),  # 23:00 NY prev day
+            'check_out': datetime(2024, 5, 30, 10, 0),  # 6:00 NY
+        })
+        # First day you only work 1 hour and second day you work 6 hours, that's -1 hours of overtime
+        self.assertAlmostEqual(early_attendance2.overtime_hours, -1, 2)
```

Comment:
is it the expected behaviour? It seems odd that the overtime is dedicated to the 30th while one hour technically fills the attendances for the 29th, no?